client/lib/tree-select/index.js

```diff
@@ -0,0 +1,69 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { isObject, some, isFunction } from 'lodash';
+
+/**
+ * Returns a selector that caches values.
+ *
+ * @param  {Function} getDependents A Function describing the dependent(s) of the selector.
+ *                                    Must return an array which gets passed as the first arg to the selector
+ * @param  {Function} selector      A standard selector for calculating cached result
+ * @return {Function}               Cached selector
+ */
+export default function treeSelect( getDependents, selector ) {
+	if ( process.env.NODE_ENV !== 'production' ) {
+		if ( ! isFunction( getDependents ) || ! isFunction( selector ) ) {
+			throw new TypeError(
+				'treeSelect: invalid arguments passed, selector and getDependents must both be functions'
+			);
+		}
+	}
+
+	const cache = new WeakMap();
+
+	const cachedSelector = function( state, ...args ) {
+		const dependents = getDependents( state, ...args );
+
+		if ( process.env.NODE_ENV !== 'production' ) {
+			if ( some( args, isObject ) ) {
+				throw new Error( 'Do not pass objects as arguments to a treeSelector' );
+			}
+		}
+
+		// create a dependency tree for caching selector results.
+		// this is beneficial over standard memoization techniques so that we can
+		// garbage collect any values that are based on outdated dependents
+		const cursor = dependents.reduce( insertDependentKey, cache );
```

Comment:
at this point I don't see much value in the `cursor` name because we're not "_cursing_" around anymore.

what about just calling it something like `leafCache` or `leaf` or `callCache` or something like that?

Reply:
i like leafCache